blockchain/utxocache.go

```diff
@@ -0,0 +1,816 @@
+	err := s.db.View(func(dbTx database.Tx) error {
+		var err error
+		entry, err = dbFetchUtxoEntryByHash(dbTx, hash)
+		return err
+	})
+
+	// Since we don't know the entries outpoint, we can't cache it.
+	return entry, err
+}
+
+// FetchEntryByHash attempts to find any available UTXO for the given hash by
+// searching the entire set of possible outputs for the given hash.
+//
+// This method is safe for concurrent access.
+func (s *utxoCache) FetchEntryByHash(hash *chainhash.Hash) (*UtxoEntry, error) {
+	s.mtx.Lock()
+	entry, err := s.getEntryByHash(hash)
+	s.mtx.Unlock()
+	return entry.Clone(), err
+}
+
+// spendEntry marks the output as spent.  Spending an output that is already
+// spent has no effect.  Entries that need not be stored anymore after being
+// spent will be removed from the cache.
+//
+// This method is part of the utxoView interface.
+// This method should be called with the state lock held.
+func (s *utxoCache) spendEntry(outpoint wire.OutPoint, addIfNil *UtxoEntry) error {
+	entry := s.cachedEntries[outpoint]
+
+	// If we don't have an entry in cache and an entry was provided, we add it.
+	if entry == nil && addIfNil != nil {
+		if err := s.addEntry(outpoint, addIfNil, false); err != nil {
+			return err
+		}
+		entry = addIfNil
+	}
+
+	// If it's nil or already spent, nothing to do.
+	if entry == nil || entry.IsSpent() {
+		return nil
+	}
+
+	// If an entry is fresh, meaning that there hasn't been a flush since it was
+	// introduced, it can simply be removed.
+	if entry.isFresh() {
+		// We don't delete it from the map, but set the value to nil, so that
+		// later lookups for the entry know that the entry does not exist in the
+		// database.
+		s.cachedEntries[outpoint] = nil
+		s.totalEntryMemory -= entry.memoryUsage()
+		return nil
+	}
+
+	// Mark the output as spent and modified.
+	entry.packedFlags |= tfSpent | tfModified
+
+	//TODO(stevenroose) check if it's ok to drop the pkScript
+	// Since we don't need it anymore, drop the pkScript value of the entry.
+	s.totalEntryMemory -= entry.memoryUsage()
+	entry.pkScript = nil
+	s.totalEntryMemory += entry.memoryUsage()
+
+	return nil
+}
+
+// addEntry adds a new unspent entry if it is not probably unspendable.  Set
+// overwrite to true to skip validity and freshness checks and simply add the
+// item, possibly overwriting another entry that is not-fully-spent.
+//
+// This method is part of the utxoView interface.
+// This method should be called with the state lock held.
+func (s *utxoCache) addEntry(outpoint wire.OutPoint, entry *UtxoEntry, overwrite bool) error {
+	// Don't add provably unspendable outputs.
+	if txscript.IsUnspendable(entry.pkScript) {
+		return nil
+	}
+
+	cachedEntry, _ := s.cachedEntries[outpoint]
+
+	// In overwrite mode, simply add the entry without doing these checks.
+	if !overwrite {
+		// Prevent overwriting not-fully-spent entries.  Note that this is not
+		// a consensus check.
+		if cachedEntry != nil && !cachedEntry.IsSpent() {
```

Comment:
The reason this code is failing at block 94804 is because there are two coinbases in the chain which share the same txid with previous coinbases and this line errors when adding the duplicate coinbase. There error bubbles up to the `commit()` function but you don't handle the error there. 

Ultimately the commit terminates and later transaction fail to validate when the utxos that failed to be committed don't exist.

Reply:
Oh god damnit Bitcoin historical bugs :D Yeah that makes sense. Any idea how Core handles this? Does it just have two if-clauses here? :/